File: leetcode/practice/Monotonic-Stack/NextGreaterElement-I.py
# 496. Next Greater Element I
# https://leetcode.com/problems/next-greater-element-i/description/

from typing import List

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        d = {}
        s = []
        for i in range(len(nums2)):
            while s and s[-1] < nums2[i]:
                a = s.pop()
                d[a] = nums2[i]
            s.append(nums2[i])
        
        return [d.get(x, -1) for x in nums1]
            

File: leetcode/practice/Monotonic-Stack/NextGreaterElement-I.js
// 496. Next Greater Element I
// https://leetcode.com/problems/next-greater-element-i/description/
var nextGreaterElement = function(nums1, nums2) {
   d = {};
   s = []; 
   nums2.forEach((item, index) => {
    while (s.length > 0 && s[s.length-1] < item) {
        a = s.pop();
        d[a] = item;
    }
    s.push(item);
   });
   return nums1.map(item => d?.[item] ?? -1);
};

console.log(nextGreaterElement([4,1,2], [1, 3, 4, 2]));